src/test/java/io/jenkins/plugins/analysis/warnings/StepsITest.java

```diff
@@ -562,6 +566,25 @@ public void shouldUseOtherJobAsReference() {
         // TODO: add verification for io.jenkins.plugins.analysis.core.model.IssueDifference
     }
 
+    /**
+     * Verifies that when publishIssues marks the build as unstable it also marks the step with
+     * WarningAction so that visualizations can display the step as unstable rather than just
+     * the whole build.
+     *
+     * @see <a href="http://issues.jenkins-ci.org/browse/JENKINS-39203">Issue 39203</a>
+     */
+    @Test
+    public void shouldMarkStepWithWarningAction() {
+        WorkflowJob job = createPipelineWithWorkspaceFiles("javac.txt");
+        job.setDefinition(asStage(createScanForIssuesStep(new Java(), "java"),
+                "publishIssues(issues:[java], qualityGates: [[threshold: 1, type: 'TOTAL', unstable: true]])"));
+        WorkflowRun run = (WorkflowRun)buildWithResult(job, Result.UNSTABLE);
+        FlowNode publishIssuesNode = new DepthFirstScanner().findFirstMatch(run.getExecution(),
+                node -> node.getDisplayFunctionName().equals("publishIssues"));
+        assertThat(publishIssuesNode).isNotNull();
+        assertThat(publishIssuesNode.getPersistentAction(WarningAction.class)).isNotNull();
```

Comment:
If we would have a subclass for `WarningAction` then we could actually check for the concrete type and see that this action if from the warnings-ng plugin.

Reply:
Yeah, we could also check that the message is the one defined in this plugin.